src/testing/initial_condition.cpp

```diff
@@ -0,0 +1,232 @@
+            // x-momentum
+            for(int d=0; d<dim; d++) {
+                gradient[d] = u*rho_grad[d] + rho*u_grad[d];
+            }
+        }
+        if(istate==2) {
+            // y-momentum
+            for(int d=0; d<dim; d++) {
+                gradient[d] = v*rho_grad[d] + rho*v_grad[d];
+            }
+        }
+        if(istate==3) {
+            // z-momentum
+            for(int d=0; d<dim; d++) {
+                gradient[d] = w*rho_grad[d] + rho*w_grad[d];
+            }
+        }
+        if(istate==4) {
+            // total energy
+            for(int d=0; d<dim; d++) {
+                gradient[d] = p_grad[d]/(1.4-1.0) + 0.5*rho_grad[d]*(u*u + v*v + w*w) + rho*(u*u_grad[d]+v*v_grad[d]+w*w_grad[d]);
+            }
+        }
+    }
+    return gradient;
+}
+
+template <int dim, typename real>
+inline dealii::Tensor<1,dim,real> InitialConditionFunction_TaylorGreenVortex<dim,real>
+::gradient (const dealii::Point<dim,real> &point, const unsigned int istate) const
+{
+    dealii::Tensor<1,dim,real> gradient = convert_primitive_to_conversative_gradient(point, istate);
+    return gradient;
+}
+//=========================================================
+// FLOW SOLVER -- Initial Condition Base Class + Factory
+//=========================================================
+template <int dim, typename real>
+InitialConditionFunction_FlowSolver<dim,real>
+::InitialConditionFunction_FlowSolver (const unsigned int nstate)
+    : dealii::Function<dim,real>(nstate)//,0.0) // 0.0 denotes initial time (t=0)
+    , nstate(nstate)
+{ 
+    // Nothing to do here yet
+}
+
+template <int dim, typename real>
+std::shared_ptr< InitialConditionFunction_FlowSolver<dim,real> > 
+InitialConditionFactory_FlowSolver<dim,real>::create_InitialConditionFunction_FlowSolver(
+    Parameters::AllParameters const *const param, 
+    int                                    nstate)
+{
+    using FlowCaseEnum = Parameters::FlowSolverParam::FlowCaseType;
+    const FlowCaseEnum flow_type = param->flow_solver_param.flow_case_type;
+
+    if(flow_type == FlowCaseEnum::inviscid_taylor_green_vortex && (dim==3)){
+        return std::make_shared < InitialConditionFunction_TaylorGreenVortex<dim,real> > (
+            nstate,
+            param->euler_param.gamma_gas,
+            param->euler_param.mach_inf);
+    }else if(flow_type == FlowCaseEnum::viscous_taylor_green_vortex && (dim==3)){
+        return std::make_shared < InitialConditionFunction_TaylorGreenVortex<dim,real> > (
+            nstate,
+            param->euler_param.gamma_gas,
+            param->euler_param.mach_inf);
+    }else{
+        std::cout << "Invalid Flow Case Type." << std::endl;
+    }
+
+    return nullptr;
+}
+
+template class InitialConditionFunction_FlowSolver <PHILIP_DIM,double>;
+template class InitialConditionFactory_FlowSolver <PHILIP_DIM,double>;
+template class InitialConditionFunction_TaylorGreenVortex <PHILIP_DIM,double>;
+
+} // PHiLiP namespace
```

Comment:
Consider moving all the initial conditions to a initial conditions subfolder in the physics folder along with FreeStream: https://github.com/dougshidong/PHiLiP/blob/040e82d3a99982dfc19ba3c6b7bb7468689f816c/src/physics/euler.h#L359-L387